### PR TITLE
[Feature] Admin approval

### LINKED ISSUE #722: Add option for new accounts to require admin approval
**Is your feature request related to a problem? Please describe.**

Right now instance owners still are lacking some measures to fight back against spam waves. One that may help is adding an option for account creation to require admin/global mod (maybe) approval before the account can be used

**Describe the solution you'd like**

We might want to investigate how other platforms do this. One major consideration is, if enabled, should accounts be made (and the address for that account name burned)? Or should the accounts be held in a separate state until admin approval, at which point they are created and the fediverse can find them.

Assuming we don't want to create the accounts in a way that uses the address forever before approval:

- Potentially a new table like `users_awaiting_approval` with columns of the `username, hashed_password, email` etc that are required to create an account
- If enabled, registrations go to this new table instead of the user table, login says awaiting approval if possible (if not, maybe user does not exist is fine at least for first pass)
- New admin UI to list entries awaiting approval
- Ability to accept or reject those requests
  - On accept, actually make the `public.user` entry, email the user
  - On deny, email the user they have been denied?

If instead we don't care if the username is burned, another possible implementation is

- New column on `public.user` such as `requires_approval = true`
- Block login if the flag is `true`
- New admin UI to list users where `requires_approval` is `true`. New index on `where requires_approval is true` to make the query fast
- Same as above on accept/deny emailing user

**Describe alternatives you've considered**

There might be other options to curtail spam that are lower hanging fruit, if people can think of them. We could also just do something like an admin option to automatically ban users on registration, so admins could then unban as a form of "approval". The flow would be confusing for users, but it's all tech that exists today so would be faster to implement

### PR DESCRIPTION
**Pull request by BentiGorlich**
- admins can now decide to turn a mandatory approval on in the admin settings
- a new tab will appear labelled "Applications" where new users that are not yet approved will appear
- when the setting is enabled a new input will be displayed on the register page: "Application Text"
- This behavior will be apparent on registering with an added flash message informing the user about it
- Users will receive an email when they are accepted or rejected
- when a user is not approved they cannot log in and an appropriate message will be displayed
- the register site now redirects to the login page instead of the front page

## Tasks

- [x] API implemented
- [x] Updated Docs
- [x] Included Images
- [x] Implemented mails
- [x] Refine email texts (I am definitely not qualified to do that)
- [x] Refine Admin application page

# Pictures

## Register Page

Register page
After registration

## Admin application view

Admin application view

## Mails

Approved email

Rejected email

Closing #722

### CODE REVIEW

**Comment by BentiGorlich** [comment]
Do we have someone that is good in phrasing user facing stuff like the emails? @jwr1 or @TheVillageGuy maybe?

File: translations/messages.en.yaml
Comment (melroy89) [inline translations/messages.en.yaml:933]:
```suggestion
email_application_rejected_body: Thank you for your interest, but we regret to inform you that your registration request has been declined.
```

File: translations/messages.en.yaml
Comment (melroy89) [inline translations/messages.en.yaml:934]:
```suggestion
email_application_pending: Your account requires admin approval before you can log in.
```

File: translations/messages.en.yaml
Comment (melroy89) [inline translations/messages.en.yaml:935]:
```suggestion
email_verification_pending: Please verify your email address before you can log in.
```

File: translations/messages.en.yaml
Comment (melroy89) [reply translations/messages.en.yaml:933]: I would call it "registration request"?

File: src/Entity/User.php
Comment (melroy89) [inline src/Entity/User.php:250]: In theory you could potentially end up with `isApproved` and `isRejected` set to true, since there is no guard against this incorrect state.

Not sure how to prevent that.

**Comment by melroy89** [comment]
The "One last step" in the message now contradict each other 😆 . Yes you still need to use the activation link, but if admin approval is required, this isn't the last step.. you than still need to wait.

**Comment by BentiGorlich** [comment]
> The "One last step" in the message now contradict each other 😆 . Yes you still need to use the activation link, but if admin approval is required, this isn't the last step.. you than still need to wait.

Yeah I know :sweat_smile:, but I didn't know how to solve this. I didn't really want to mess up the existing translations... But if you think that it is necessary I will split the existing one up, so that both messages can be merged

File: src/Entity/User.php
Comment (melroy89) [reply src/Entity/User.php:250]: Ah yes of course, this needs to be an enum type. This will also allows you to introduce other states like "waiting" (waiting for approval) safely. 

This is the way.

**Comment by BentiGorlich** [comment]
@melroy89 are you fine with the 2 flash messages upon signup? I think it is fine this way, but I can invest the work to fix it if you want to

**Comment by melroy89** [comment]
> @melroy89 are you fine with the 2 flash messages upon signup? I think it is fine this way, but I can invest the work to fix it if you want to

It's fine for me.

File: src/Entity/User.php
Comment (melroy89) [reply src/Entity/User.php]: Seems to be fixed now, right? Looks much better with an enum 🥳

**Review by melroy89 — APPROVED** [review]

**Pull request merged**
